code/game/objects/items/devices/gps.dm

```diff
@@ -1,28 +1,423 @@
+				else
+					dat += "<td>([gps["x"]], [gps["y"]], [gps["z_name"]])</td>"
+
+				if(gps["local"])
+					dat += "<td>[gps["distance"]]m</td><td>[gps["direction"]]</td>"
+				else
+					dat += "<td colspan = 2>Non-local signal.</td>"
+
+				if(LAZYACCESS(tracking_devices, gps_ref))
+					dat += "<td><a href='?src=\ref[src];stop_track=[gps_ref]'>Stop Tracking</a> <a href='?src=\ref[src];track_color=[gps_ref]'>Colour [COLORED_SQUARE(LAZYACCESS(tracking_devices, gps_ref))]</a> <a href='?src=\ref[src];track_label=[gps_ref]'>Show/Hide Label</a></td>"
+				else
+					dat += "<td><a href='?src=\ref[src];start_track=[gps_ref]'>Start Tracking</a></td>"
+				dat += "</tr>"
+		else
+			dat += "<tr><td colspan = 5>No other signals detected.</td></tr>"
+
+	dat += "</table>"
+
+	var/datum/browser/popup = new(user, "gps_\ref[src]", "Global Positioning System", 700, 1000)
+	popup.set_content(dat.Join(null))
+	popup.open()
+
+/obj/item/gps/OnTopic(var/mob/user, var/list/href_list)
+
+	if(href_list["toggle_power"])
+		toggle_tracking()
+		. = TOPIC_HANDLED
+
+	if(href_list["track_label"])
+		var/gps_ref = href_list["track_label"]
+		var/obj/item/gps/gps = locate(gps_ref)
+		if(istype(gps) && !QDELETED(gps) && !LAZYACCESS(showing_tracked_names, gps_ref))
+			LAZYSET(showing_tracked_names, gps_ref, TRUE)
+		else
+			LAZYREMOVE(showing_tracked_names, gps_ref)
+		to_chat(user, SPAN_NOTICE("\The [src] is [LAZYACCESS(showing_tracked_names, gps_ref) ? "now showing" : "no longer showing"] labels for [gps.gps_tag]."))
+		. = TOPIC_HANDLED
+
+	if(href_list["stop_track"])
+		var/gps_ref = href_list["stop_track"]
+		var/obj/item/gps/gps = locate(gps_ref)
+		compass.clear_waypoint(gps_ref)
+		LAZYREMOVE(tracking_devices, gps_ref)
+		LAZYREMOVE(showing_tracked_names, gps_ref)
+		if(istype(gps) && !QDELETED(gps))
+			to_chat(user, SPAN_NOTICE("\The [src] is no longer tracking [gps.gps_tag]."))
+		update_compass()
+		. = TOPIC_HANDLED
+
+	if(href_list["start_track"])
+		var/gps_ref = href_list["start_track"]
+		var/obj/item/gps/gps = locate(gps_ref)
+		if(istype(gps) && !QDELETED(gps))
+			LAZYSET(tracking_devices, gps_ref, COLOR_SILVER)
+			LAZYSET(showing_tracked_names, gps_ref, TRUE)
+			to_chat(user, SPAN_NOTICE("\The [src] is now tracking [gps.gps_tag]."))
+			update_compass()
+			. = TOPIC_HANDLED
+
+	if(href_list["track_color"])
+		var/obj/item/gps/gps = locate(href_list["track_color"])
+		if(istype(gps) && !QDELETED(gps))
+			var/new_colour = input("Enter a new tracking color.", "GPS Waypoint Color") as color|null
+			if(new_colour && istype(gps) && !QDELETED(gps) && holder == user && !user.incapacitated())
+				to_chat(user, SPAN_NOTICE("You adjust the colour \the [src] is using to highlight [gps.gps_tag]."))
+				LAZYSET(tracking_devices, href_list["track_color"], new_colour)
+				update_compass()
+				. = TOPIC_HANDLED
+
+	if(href_list["tag"])
+		var/a = input("Please enter desired tag.", name, gps_tag) as text
+		a = uppertext(copytext(sanitize(a), 1, 11))
```

Comment:
Are you sure you want to sanitize here? The main thing this proc will do is html-encode & and similar characters. If you _do_ in fact want that, you may need to decode in the default of `input` above, though as you are also truncating it won't always work.

Reply:
I'm a bit confused by this comment, I was under the impression it was standard practice to sanitize after an `as text` input. This is meant to be a very short all caps string so I'm not too fussed if people want to put weird stuff into it and get nonsense as their default.